### PR TITLE
BLD: Only enable -Werror in the CI jobs

### LINKED ISSUE #33314: BUG: clang-9: error: -Wl,-export_dynamic: 'linker' input unused (OSX)
- [x] I have checked that this issue has not already been reported.

- [ ] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

### conda environment

```
conda create -n pandas-compile-test compilers numpy cython
```

<details>

```

  binutils-meta      conda-forge/osx-64::binutils-meta-1.0.4-0
  c-compiler         conda-forge/osx-64::c-compiler-1.0.4-h1239861_0
  ca-certificates    conda-forge/osx-64::ca-certificates-2020.4.5.1-hecc5488_0
  cctools            conda-forge/osx-64::cctools-927.0.2-h5ba7a2e_4
  certifi            conda-forge/osx-64::certifi-2020.4.5.1-py38h32f6830_0
  clang              conda-forge/osx-64::clang-9.0.1-default_hf57f61e_0
  clang_osx-64       conda-forge/osx-64::clang_osx-64-9.0.1-h05bbb7f_0
  clangxx            conda-forge/osx-64::clangxx-9.0.1-default_hf57f61e_0
  clangxx_osx-64     conda-forge/osx-64::clangxx_osx-64-9.0.1-h05bbb7f_0
  compiler-rt        conda-forge/osx-64::compiler-rt-9.0.1-h6a512c6_3
  compiler-rt_osx-64 conda-forge/noarch::compiler-rt_osx-64-9.0.1-h99342c6_3
  compilers          conda-forge/osx-64::compilers-1.0.4-0
  cxx-compiler       conda-forge/osx-64::cxx-compiler-1.0.4-h707564b_0
  cython             conda-forge/osx-64::cython-0.29.16-py38hc84c608_0
  fortran-compiler   conda-forge/osx-64::fortran-compiler-1.0.4-he991be0_0
  gfortran_impl_osx~ conda-forge/osx-64::gfortran_impl_osx-64-7.3.0-hf4212f2_2
  gfortran_osx-64    conda-forge/osx-64::gfortran_osx-64-7.3.0-h22b1bf0_9
  gmp                conda-forge/osx-64::gmp-6.2.0-h4a8c4bd_2
  isl                conda-forge/osx-64::isl-0.19-0
  ld64               conda-forge/osx-64::ld64-450.3-h3c32e8a_4
  libblas            conda-forge/osx-64::libblas-3.8.0-16_openblas
  libcblas           conda-forge/osx-64::libcblas-3.8.0-16_openblas
  libcxx             conda-forge/osx-64::libcxx-9.0.1-1
  libffi             conda-forge/osx-64::libffi-3.2.1-h4a8c4bd_1007
  libgfortran        conda-forge/osx-64::libgfortran-4.0.0-2
  libiconv           conda-forge/osx-64::libiconv-1.15-h0b31af3_1006
  liblapack          conda-forge/osx-64::liblapack-3.8.0-16_openblas
  libllvm9           conda-forge/osx-64::libllvm9-9.0.1-ha1b3eb9_0
  libopenblas        conda-forge/osx-64::libopenblas-0.3.9-h3d69b6c_0
  llvm-openmp        conda-forge/osx-64::llvm-openmp-9.0.1-h28b9765_2
  mpc                conda-forge/osx-64::mpc-1.1.0-h4160ff4_1006
  mpfr               conda-forge/osx-64::mpfr-4.0.2-h44b798e_0
  ncurses            conda-forge/osx-64::ncurses-6.1-h0a44026_1002
  numpy              conda-forge/osx-64::numpy-1.18.1-py38h1f821a2_1
  openssl            conda-forge/osx-64::openssl-1.1.1f-h0b31af3_0
  pip                conda-forge/noarch::pip-20.0.2-py_2
  python             conda-forge/osx-64::python-3.8.2-hd5f0129_5_cpython
  python_abi         conda-forge/osx-64::python_abi-3.8-1_cp38
  readline           conda-forge/osx-64::readline-8.0-hcfe32e1_0
  setuptools         conda-forge/osx-64::setuptools-46.1.3-py38h32f6830_0
  sqlite             conda-forge/osx-64::sqlite-3.30.1-h93121df_0
  tapi               conda-forge/osx-64::tapi-1000.10.8-ha1b3eb9_4
  tk                 conda-forge/osx-64::tk-8.6.10-hbbe82c9_0
  wheel              conda-forge/noarch::wheel-0.34.2-py_1
  xz                 conda-forge/osx-64::xz-5.2.4-h0b31af3_1002
  zlib               conda-forge/osx-64::zlib-1.2.11-h0b31af3_1006
```
</details>

Error message:

```
clang-9: error: -Wl,-export_dynamic: 'linker' input unused [-Werror,-Wunused-command-line-argument]
```

Output of `pip install -e ~/Development/pandas --no-deps --no-use-pep517 -vvv`:

<details>

```
Non-user install because site-packages writeableCreated temporary directory: /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-ephem-wheel-cache-ehxkyhsg
Created temporary directory: /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv
Initialized build tracking at /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv
Created build tracker: /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv
Entered build tracker: /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv
Created temporary directory: /private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-install-5lvoa7_h
Obtaining file:///Users/uwe/Development/pandas
  Added file:///Users/uwe/Development/pandas to build tracker '/private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv'
    Running setup.py (path:/Users/uwe/Development/pandas/setup.py) egg_info for package from file:///Users/uwe/Development/pandas
    Running command python setup.py egg_info
    Compiling pandas/_libs/hashing.pyx because it depends on /Users/uwe/miniconda3/envs/pandas-compile-test/lib/python3.8/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling pandas/_libs/reshape.pyx because it depends on /Users/uwe/miniconda3/envs/pandas-compile-test/lib/python3.8/site-packages/Cython/Includes/numpy/__init__.pxd.
    Compiling pandas/_libs/window/indexers.pyx because it depends on /Users/uwe/miniconda3/envs/pandas-compile-test/lib/python3.8/site-packages/Cython/Includes/numpy/__init__.pxd.
    [1/3] Cythonizing pandas/_libs/hashing.pyx
    [2/3] Cythonizing pandas/_libs/reshape.pyx
    [3/3] Cythonizing pandas/_libs/window/indexers.pyx
    running egg_info
    writing pandas.egg-info/PKG-INFO
    writing dependency_links to pandas.egg-info/dependency_links.txt
    writing entry points to pandas.egg-info/entry_points.txt
    writing requirements to pandas.egg-info/requires.txt
    writing top-level names to pandas.egg-info/top_level.txt
    reading manifest file 'pandas.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    no previously-included directories found matching 'doc/build'
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '*~' found anywhere in distribution
    warning: no previously-included files matching '.DS_Store' found anywhere in distribution
    warning: no previously-included files matching '#*' found anywhere in distribution
    writing manifest file 'pandas.egg-info/SOURCES.txt'
  Source in /Users/uwe/Development/pandas has version 1.1.0.dev0+1122.g01f73100d.dirty, which satisfies requirement pandas==1.1.0.dev0+1122.g01f73100d.dirty from file:///Users/uwe/Development/pandas
  Removed pandas==1.1.0.dev0+1122.g01f73100d.dirty from file:///Users/uwe/Development/pandas from build tracker '/private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv'
Installing collected packages: pandas
  Running setup.py develop for pandas
    Running command /Users/uwe/miniconda3/envs/pandas-compile-test/bin/python3.8 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/Users/uwe/Development/pandas/setup.py'"'"'; __file__='"'"'/Users/uwe/Development/pandas/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
    running develop
    running egg_info
    writing pandas.egg-info/PKG-INFO
    writing dependency_links to pandas.egg-info/dependency_links.txt
    writing entry points to pandas.egg-info/entry_points.txt
    writing requirements to pandas.egg-info/requires.txt
    writing top-level names to pandas.egg-info/top_level.txt
    reading manifest file 'pandas.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    no previously-included directories found matching 'doc/build'
    warning: no previously-included files matching '*.pyd' found anywhere in distribution
    warning: no previously-included files matching '*~' found anywhere in distribution
    warning: no previously-included files matching '.DS_Store' found anywhere in distribution
    warning: no previously-included files matching '#*' found anywhere in distribution
    writing manifest file 'pandas.egg-info/SOURCES.txt'
    running build_ext
    building 'pandas._libs.algos' extension
    x86_64-apple-darwin13.4.0-clang -fno-strict-aliasing -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O3 -Wall -Wstrict-prototypes -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O3 -pipe -fdebug-prefix-map=${SRC_DIR}=/usr/local/src/conda/${PKG_NAME}-${PKG_VERSION} -fdebug-prefix-map=/Users/uwe/miniconda3/envs/pandas-compile-test=/usr/local/src/conda-prefix -flto -Wl,-export_dynamic -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O3 -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include -D_FORTIFY_SOURCE=2 -mmacosx-version-min=10.9 -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include -DNPY_NO_DEPRECATED_API=0 -I./pandas/_libs -Ipandas/_libs/src/klib -I/Users/uwe/miniconda3/envs/pandas-compile-test/lib/python3.8/site-packages/numpy/core/include -I/Users/uwe/miniconda3/envs/pandas-compile-test/include/python3.8 -c pandas/_libs/algos.c -o build/temp.macosx-10.9-x86_64-3.8/pandas/_libs/algos.o -Werror
    clang-9: error: -Wl,-export_dynamic: 'linker' input unused [-Werror,-Wunused-command-line-argument]
    error: command 'x86_64-apple-darwin13.4.0-clang' failed with exit status 1
Cleaning up...
Removed build tracker: '/private/var/folders/3r/k0d_q3hs2mb2_jxfh_9hjyqr0000gn/T/pip-req-tracker-waeimcjv'
ERROR: Command errored out with exit status 1: /Users/uwe/miniconda3/envs/pandas-compile-test/bin/python3.8 -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/Users/uwe/Development/pandas/setup.py'"'"'; __file__='"'"'/Users/uwe/Development/pandas/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

</details>

`export | grep FLAGS`:
 
```
CFLAGS='-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
CONDA_BACKUP_LDFLAGS=' -Wl,-rpath,/Users/uwe/miniconda3/envs/pandas-compile-test/lib -L/Users/uwe/miniconda3/envs/pandas-compile-test/lib'
CPPFLAGS='-D_FORTIFY_SOURCE=2 -mmacosx-version-min=10.9 -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
CXXFLAGS='-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
DEBUG_CFLAGS='-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -Og -g -Wall -Wextra -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
DEBUG_CXXFLAGS='-march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -Og -g -Wall -Wextra -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
DEBUG_FFLAGS='-march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include -march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -Og -g -Wall -Wextra -fcheck=all -fbacktrace -fimplicit-none -fvar-tracking-assignments'
DEBUG_FORTRANFLAGS='-march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include -march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -Og -g -Wall -Wextra -fcheck=all -fbacktrace -fimplicit-none -fvar-tracking-assignments'
FFLAGS='-march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
FORTRANFLAGS='-march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -isystem /Users/uwe/miniconda3/envs/pandas-compile-test/include'
LDFLAGS=' -Wl,-rpath,/Users/uwe/miniconda3/envs/pandas-compile-test/lib -L/Users/uwe/miniconda3/envs/pandas-compile-test/lib'
LDFLAGS_LD='-pie -headerpad_max_install_names -dead_strip_dylibs -rpath /Users/uwe/miniconda3/envs/pandas-compile-test/lib -L/Users/uwe/miniconda3/envs/pandas-compile-test/lib'
XPC_FLAGS=0x0
```

Seems like this is not the first issue in response to https://github.com/pandas-dev/pandas/pull/32163, there is also e.g. https://github.com/pandas-dev/pandas/issues/33224

### LINKED ISSUE #33315: Remove `-Werror` from setup.py
It seems nice to have `-Werror` in CI but as a conda-forge maintainer I quite often have to write patches to remove `-Werror` from releases as the system distributions build packages on are not identical to the ones CI of upstream runs on. There is a large variety of compiler warnings and some may only show up when other C-defines are set. Also many new warning appear when you build with a newer compiler or different versions of the dependencies used.

My suggestion would be:
 * Remove `-Werror` from `setup.py`.
 * Add `-Werror` to `CFLAGS` on CI to have the same result for pandas-dev as it is currently.

Original PR: https://github.com/pandas-dev/pandas/pull/32163

Related issues: https://github.com/pandas-dev/pandas/issues/33314, https://github.com/pandas-dev/pandas/issues/33224

### PR DESCRIPTION
**Pull request by xhochy**
closes #33315
closes #33314
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by xhochy** [comment]
In which `whatsnew` section should this go? Bug fixes - other?

**Comment by jorisvandenbossche** [comment]
Thanks! "Bug fixes - other" is fine for me 
(in pandas 1.0.0, we had a "build changes" section (https://pandas.pydata.org/docs/whatsnew/v1.0.0.html#build-changes), but that was a more relevant change, related to pyproject.toml)

File: setup.py [hunk 1]
```diff
@@ -409,15 +409,15 @@ def run(self):
...
 if is_platform_windows():
-    extra_compile_args = []
-    extra_link_args = []
     if debugging_symbols_requested:
         extra_compile_args.append("/Z7")
         extra_link_args.append("/DEBUG")
 else:
-    extra_compile_args = ["-Werror"]
-    extra_link_args = []
+    if os.environ.get("PANDAS_CI", "0") == "1":
```
Comment (jreback) [inline setup.py:419]: can you add a comment that this is set explicity in ci/setup_env.sh

**Review by jreback — CHANGES_REQUESTED** [review]
very minor comment, lgtm

**Comment by jreback** [comment]
also not averse to do this for 1.1.5 as well.

**Comment by simonjayhawkins** [comment]
> also not averse to do this for 1.1.5 as well.

sure. however not an issue on MacPython for 1.1.5. Is causing issues with 1.2 though. see https://github.com/pandas-dev/pandas/issues/37784#issuecomment-732911912

https://github.com/MacPython/pandas-wheels/pull/112 my not be needed with this change.

**Comment by jreback** [comment]
> > also not averse to do this for 1.1.5 as well.
> 
> sure. however not an issue on MacPython for 1.1.5. Is causing issues with 1.2 though. see [#37784 (comment)](https://github.com/pandas-dev/pandas/issues/37784#issuecomment-732911912)
> 
> [MacPython/pandas-wheels#112](https://github.com/MacPython/pandas-wheels/pull/112) my not be needed with this change.

interesting ok, then fine not to backport.

**Comment by simonjayhawkins** [comment]
> interesting ok, then fine not to backport.

it's not causing us issues, but i'm ok with backporting if it could be beneficial to others

**Comment by xhochy** [comment]
> > interesting ok, then fine not to backport.
> 
> it's not causing us issues, but i'm ok with backporting if it could be beneficial to others

It would be helpful for the conda packages on macOS

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -766,6 +766,7 @@ Other
 - Bug in :meth:`Index.union` behaving differently depending on whether operand is an :class:`Index` or other list-like (:issue:`36384`)
 - Passing an array with 2 or more dimensions to the :class:`Series` constructor now raises the more specific ``ValueError`` rather than a bare ``Exception`` (:issue:`35744`)
 - Bug in ``dir`` where ``dir(obj)`` wouldn't show attributes defined on the instance for pandas objects (:issue:`37173`)
+- Only set ``-Werror`` as a compiler flag in the CI jobs (:issue:`33315`, :issue:`33314`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:769]: ok happy to backport, can you move to 1.1.5 whatnew

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v1.2.0.rst (hunk 2 above)
Comment (jreback) [reply doc/source/whatsnew/v1.2.0.rst:769]: if you'd move the note, pls ping on green (might be a while, merged a lot of things today :->)

**Comment by simonjayhawkins** [comment]
all green

**Comment by jreback** [comment]
this needs the note moved to 1.1.5

**Comment by simonjayhawkins** [comment]
test failure unrelated

```
=========================== short test summary info ============================
FAILED pandas/tests/plotting/test_misc.py::TestDataFramePlots::test_scatter_matrix_axis
= 1 failed, 143660 passed, 3518 skipped, 1408 xfailed, 14 warnings in 1846.77s (0:30:46) =
```

**Comment by jorisvandenbossche** [comment]
Thanks @xhochy !

**Comment by simonjayhawkins** [comment]
@meeseeksdev backport 1.1.x

**Pull request merged**